core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala

```diff
@@ -51,27 +54,46 @@ class LogDirFailureTest extends IntegrationTestHarness {
   @Before
   override def setUp() {
     super.setUp()
-    TestUtils.createTopic(zkUtils, topic, 1, 2, servers = servers)
+    TestUtils.createTopic(zkUtils, topic, partitionNum, serverCount, servers = servers)
   }
 
   @Test
   def testIOExceptionDuringLogRoll() {
-    testProduceAfterLogDirFailure(Roll)
+    testProduceAfterLogDirFailureOnLeader(Roll)
   }
 
   @Test
   def testIOExceptionDuringCheckpoint() {
-    testProduceAfterLogDirFailure(Checkpoint)
+    testProduceAfterLogDirFailureOnLeader(Checkpoint)
+  }
+
+  @Test
+  def testReplicaFetcherThreadAfterLogDirFailureOnFollower() {
+    val consumer = consumers.head
+    subscribeAndWaitForAssignment(topic, consumer)
+    val producer = producers.head
+    val partition = new TopicPartition(topic, 0)
+
+    val partitionInfo = producer.partitionsFor(topic).asScala.find(_.partition() == 0).get
+    val leaderServerId = partitionInfo.leader().id()
+    val followerId = partitionInfo.replicas().map(_.id()).find(_ != leaderServerId).get
+    val followerServer = servers.find(_.config.brokerId == followerId).get
+
+    followerServer.replicaManager.markPartitionOffline(partition)
+    TestUtils.produceMessage(servers, topic, 0, "message")
```

Comment:
Do we really to produce any message for the test? It seems that we need to make sure that ReplicaFetcherThread has called buildFetchRequest() at least once after markPartitionOffline() is called. It doesn't seem that the test guarantees that?

Reply:
Currently a FetchRequest will be put in the purgatory for up to 500 ms if there is no data. Thus ReplicaFetcherThread will wait up to 500 ms after log dir failure before it receives the FetchResponse and triggers the bug. If we don't produce message, then the last AssertTrue() may fail if assertion is made before ReplicaFetcherThread receives response. Alternatively we can use e.g. waitUntilTrue but that is a bit slower than producing a message.

The bug (i.e. None.get) will be triggered when ReplicaFetcherThread.processPartitionData() is called. After we produce the message in sync-mode with acks = -1, it is guaranteed that the follower should have fetched this message from leader. Which means follower has called ReplicaFetcherThread.processPartitionData() after the partition is marked offline. Then None.get() will triggered if the bug exists. Does this make sense?

Reply:
I have updated the test such that the test will send a message to another partition which has the same leader as the offline partition. By sending this message after we mark the replica offline, we make sure that ReplicaFetcherThread will get the FetchResponse from the leader immediately after the replica is offline, instead of having the FetchRequest wait in the purgatory.

Also note that by sending a message with acks = -1, by the time producer.send(record).get() returns, it is guaranteed that the ReplicaFetcherThread on the follower has already called processPartitionData() on the offline replica and then sent another FetchResponse to the leader. This means the bug (if still exists) is guaranteed to be triggered.

Reply:
Well, a couple of considerations here. (1) It's possible that when produceMessage() returns, isr is shrunk to 1. In this case, it's not clear if a fetch request from the follower has been issued. This is rare, but in theory can happen. (2) In the more common case, after markPartitionOffline(), the follower will be excluding the partition in the fetch request. This will cause produceMessage() to wait for replica.max.lag.ms to drop the follower out of isr before acking the producer. This will mean a 10 sec delay in the test, which is not ideal.

I am just wonder if there are simpler ways to test this. For example, I was thinking that we could introduce a counter in AbstractFetcherThread that increments on each fetch request. Then, we can just wait until the counter increments after the partition is marked offline and then check the thread is still alive.

Reply:
@junrao Regarding 1), note that the test will produce one message to a partition that is not marked offline on the follower. Thus isr for this partition should be 2 because follower will keep fetching from the leader after partition 0 is marked offline. I am not sure why isr of this partition will drop to 1. I have updated the test to assert that isr is 2 at the end of the test.

Regarding 2), because the test is producing message to a partition that is not marked offline, the produce should succeed immediately without waiting for 10 seconds.  I also verified that producer.send().get takes only 170 ms by adding log statement in my test.

Regarding your suggestion, that is good point. There is indeed a simpler way to verify this using lastCaughtUpTimeMs which we already maintain in Replica. I have updated the patch to use this without having a producer sends message. Can you see if it looks OK now?

Reply:
@junrao Sorry for being back-and-forth on this. I realized that the original approach of sending messages with ack = -1 is better than waiting for timestamp to increase. The reason is that, after we mark the partition offline on the follower, we need to wait for follower to fetch from leader and try to append to the local replica, before we verify that the replica fetcher thread is alive. 

This will be guaranteed when producer.send() returns. But this is not guaranteed if only check the lastCaughtUpTimeMs -- it is possible that lastCaughtUpTimeMs has increased but replica fetcher thread has not attempted to append to the local replica.

So I reverted to the approach as of this comment and added an assertion to make sure isr == serverCount. Can you see if it is OK?